### PR TITLE
Do not fail builds with absolute paths in cSources

### LINKED ISSUE #11998: `Cabal internal error: makeRelativePathEx` in `Link.hs`
**Describe the bug**
When a Cabal module generates a C source file under the `dist-newstyle` tree, appends it to `cSources`, and `dist-newstyle` directory is not a descendant of the Cabal module directory, then the build will fail with the following error:

```
Failed to build setup-hooks-non-hs-nested-0.1.0.0-inplace. The exception was:
  |   Cabal internal error: makeRelativePathEx: absolute path /tmp/cabal-testsuite-559760/cabal.dist/work/./dist/build/x86_64-linux/ghc-9.14.1/setup-hooks-non-hs-nested-0.1.0.0/build/inner/autogen/Gen2.o

HasCallStack backtrace:
  error, called at src/Distribution/Utils/Path.hs:257:7 in Cabal-syntax-3.17.0.0-e21a35722f0b73c9611db8cb96bb900ae64ee5fe20bbad74d964fb424d28ba46:Distribution.Utils.Path
  makeRelativePathEx, called at src/Distribution/Simple/GHC/Build/Link.hs:114:11 in Cabal-3.17.0.0-5a30e95397b38fc4f23d484a1dfa38a5f19c27899cfa21eb728fbabfee76161c:Distribution.Simple.GHC.Build.Link
```

**To Reproduce**
I have a `cabal-testsuite` reproducer here:
https://github.com/e-rk/cabal/tree/cabal-nested-build-test-cSources/cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested

It is basically a copy of `SetupHooksNonHs`, except that the Cabal module is nested in a subdirectory relative to `cabal build` workdir.

**Expected behavior**
The build should succeed regardless of where `dist-newstyle` is located relative to the directory containing the Cabal module.

**System information**
 - Operating system: Archlinux, Windows 11
 - `cabal`: `master` branch
 - `ghc`: 9.14.1

**Additional context**
The error points to: https://github.com/haskell/cabal/blob/dc2d83118b2dc410fb9fd6d18c00f13b1139449c/Cabal/src/Distribution/Simple/GHC/Build/Link.hs#L114
From what I see, when `cabal build` is invoked from the parent directory of a Cabal module, then API such as `buildDir lbi` will return an absolute path. 

This is in contrast to when the build is invoked directly in the Cabal module directory, in which case the `buildDir lbi` will be relative and will not trigger this error. 

It's very likely that all path functions that read `CommonBuildFlags setupDistPref` underneath have the same behavior.

### PR DESCRIPTION
**Pull request by e-rk**
When a build hook supplied a cSources path that was absolute, the linking stage of the build would fail as it expected the paths to be always relative.

This poses a problem for Cabal modules that generate C source files at build time. The reason is that when dist-newstyle is located outside the Cabal module tree, such as when it is built as part of a larger cabal.project, the path APIs, such as:
 - autogenComponentModulesDir
 - buildDir
 - ...and so on

will produce absolute paths.

Normally those APIs produce relative paths when the build directory is located somewhere under the module tree, and therefore don't cause build issues.

It is easy for the developer to believe that they should use the above functions to get the path to the build/autogen directory, and only then realize that build completely fails apart at linking stage due to presence of absolute paths under certain circumstances.

The change allows the Cabal to process absolute paths in cSources, so the paths produced by above APIs can always be used safely.

Added a test that verifies this build scenario is covered.

Added changelog entry.

Fixes #11998

---

**Template Α: This PR modifies [behaviour or interface](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog)**

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
  * [x] [Is the change significant?](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#is-my-change-significant) If so, remember to add `significance: significant` in the changelog file.
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

### CODE REVIEW

File: cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal [hunk 1]
```diff
@@ -0,0 +1,38 @@
+cabal-version:       3.14
+name:                setup-hooks-non-hs-nested
```
Comment (sheaf) [inline cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal]: I would much prefer if you updated the existing `setup-hooks-non-hs` test to use a combination of absolute and relative paths, as opposed to duplicating the entire test.

File: Cabal/src/Distribution/Simple/GHC/Build/Link.hs
Comment (sheaf) [inline Cabal/src/Distribution/Simple/GHC/Build/Link.hs]: Can we make this into a separate top-level function and re-use it instead?

Here's a suggestion which also documents a bit more clearly what happens with absolute paths:


```hs
-- | Given the path to a source file (Haskell, C, JS...), return the corresponding
-- object file path (input to the linker).
objectFilePath
  :: SymbolicPath Pkg (Dir Artifacts)
  -- ^ objects build directory (@-odir@)
  -> String
  -- ^ object file extension
  -> SymbolicPath Pkg File
  -- ^ source file path (relative or absolute)
  -> SymbolicPath Pkg File
  -- ^ resulting object file path
objectFilePath buildTargetDir objExt src =
  case symbolicPathRelative_maybe objPath of
    Just relObjPath ->
      -- For a relative input file, the output object file is placed under the -odir.
      coerceSymbolicPath buildTargetDir </> relObjPath
    Nothing ->
      -- For an absolute input file, the object file is placed next to the input file
      -- and not within -odir.
      objPath
  where
    objPath = src `replaceExtensionSymbolicPath` objExt
```

File: changelog.d/issue-11998.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+synopsis: Fixed build failure with absolute paths in `BuildInfo` `cSources`
+packages: Cabal
+prs: #0000
```
Comment (sheaf) [inline changelog.d/issue-11998.md]:
```suggestion
prs: #12005
```

**Review by sheaf — COMMENTED** [review]
The fix looks great. I just made some suggestions to reduce code duplication and improve the documentation.

File: Cabal/src/Distribution/Simple/GHC/Build/Link.hs
Comment (e-rk) [reply Cabal/src/Distribution/Simple/GHC/Build/Link.hs]: Done as suggested. Thank you for the suggestion.
The function was put into `Distribution.Simple.GHC.Build.Utils`. In that same file there is also `getObjectFileName`, which appears to have similar function, but takes different parameters. I was reluctant to remove it and use `objectFilePath`, however I have modified to use `objectFilePath` internally.

File: changelog.d/issue-11998.md (hunk 2 above)
Comment (e-rk) [reply changelog.d/issue-11998.md]: Done. Also added one more sentence explaining that `autogenComponentModulesDir` should be used to obtain the `autogen` directory path.

File: cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal (hunk 1 above)
Comment (e-rk) [reply cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal]: Added `inner` directory with only the `cabal.project` file to ensure that both in-tree and out-of-tree build scenarios have coverage with just a single test implementation.

My initial attempt at this was adding a rule that always uses an absolute path to generated C source file regardless of how the module is being built.
I have such change on a separate branch here: 
https://github.com/e-rk/cabal/blob/9df2bc1839d94038273d620436c7c914cd82622c/cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHs/SetupHooks.hs#L178-L182

I haven't added it to this PR, because I encountered a different set of problems in that test scenario.
The `SetupHooks/Internal.hs` build graph implementation might have an unexpected behavior with absolute `cSources` paths when the `autogenComponentModulesDir` is relative, but it might have also been a mistake in my test implementation.
I will investigate this further, but I think such test could be added on a separate occasion in a different PR. The question is also whether such scenario is an edge condition that the maintainers consider not worth covering in the first place.

**Comment by e-rk** [comment]
Thank you for the review.
Pushed a commit addressing review comments.
I don't know if force-pushes are welcome during reviews, so I made a fixup commit that I will squash later on before merging.

File: Cabal/src/Distribution/Simple/GHC/Build/Utils.hs [hunk 3]
```diff
@@ -106,17 +105,44 @@ checkNeedsRecompilation mbWorkDir filename opts =
 -- | Finds the object file name of the given source file
 getObjectFileName
   :: Maybe (SymbolicPath CWD (Dir Pkg))
+  -- ^ Package build directory
```
Comment (sheaf) [inline Cabal/src/Distribution/Simple/GHC/Build/Utils.hs]:
```suggestion
  -- ^ Package directory
```

**Review by sheaf — APPROVED** [review]

File: Cabal/src/Distribution/Simple/GHC/Build/Utils.hs (hunk 3 above)
Comment (e-rk) [reply Cabal/src/Distribution/Simple/GHC/Build/Utils.hs]: Done

**Comment by e-rk** [comment]
Squashed the fixup commits.

File: cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal (hunk 1 above)
Comment (sheaf) [reply cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal]: Hmm, I don't completely follow what the issue is. Could you possibly open a separate ticket about what the issue is? I'd be happy to investigate.

**Review by zlonast — APPROVED** [review]

File: cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal (hunk 1 above)
Comment (ulysses4ever) [reply cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal]: We need to close all conversations in order to merge

Comment (ulysses4ever) [reply cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal]: Please, do it when you feel it's a good time

Comment (e-rk) [reply cabal-testsuite/PackageTests/SetupHooks/SetupHooksNonHsNested/inner/setup-hooks-non-hs-nested.cabal]: Sorry for not responding! I did not have the time to investigate the issue I was talking about earlier. But it is something separate and shouldn't be a blocker for merging this PR.

**Comment by Mikolaj** [comment]
Backporting due to https://github.com/haskell/cabal/pull/12054#issuecomment-4881229225

**Comment by Mikolaj** [comment]
@mergify backport 3.18
